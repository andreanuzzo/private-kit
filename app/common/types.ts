import { Posix } from '../helpers/dateTimeUtils';

export type LocalDataPoint = {
  latitude: number;
  longitude: number;
  time: Posix;
  hashes: string[];
};

export type Coordinates = {
  latitude: number;
  longitude: number;
};

export type ApiHealthcareAuthority = {
  name: string;
  bounds: {
    ne: Coordinates;
    sw: Coordinates;
  };
  org_id: string;
  cursor_url: string;
  public_api: string;
};

export type HealthcareAuthority = ApiHealthcareAuthority & {
  internal_id: string;
};
