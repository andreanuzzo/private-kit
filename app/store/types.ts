import { store } from './index';
import rootReducer from './reducers/rootReducer';
import { StateType } from 'typesafe-actions';
import { ThunkAction } from 'redux-thunk';
import { Action } from 'redux';
// Our internal store representation is the same as the API representation
export type { HealthcareAuthority } from '../common/types';

export type RootState = ReturnType<typeof rootReducer>;
export type Store = StateType<typeof store>;

// These will appear in redux debugger, so string names are helpful
export enum ApiStatus {
  INITIAL = 'INITIAL',
  STARTED = 'STARTED',
  SUCCESS = 'SUCCESS',
  FAILURE = 'FAILURE',
}

export type ApiRequest = {
  status: ApiStatus; // This allows components to show states based on api status
  errorMessage: string | null; // This is only for redux debugging. Store as a string for the safety in the store.
};

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export enum FeatureFlagOption {
  IMPORT_LOCATIONS_GOOGLE = 'IMPORT_LOCATIONS_GOOGLE',
  IMPORT_LOCATIONS_JSON_URL = 'IMPORT_LOCATIONS_JSON_URL',
  CUSTOM_URL = 'CUSTOM_URL',
  DOWNLOAD_LOCALLY = 'DOWNLOAD_LOCALLY',
  DEV_LANGUAGES = 'DEV_LANGUAGES',
  BYPASS_EXPORT_API = 'BYPASS_EXPORT_API',
  MOCK_EXPOSURE = 'MOCK_EXPOSURE',
}
