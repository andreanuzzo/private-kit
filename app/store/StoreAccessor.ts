import { Store } from 'redux';

/*
 * This accessor is added for the SOLE purpose of avoiding cyclic imports from
 * the IntersectService to the Intersect helper and again to the store. The
 * accessor serves as a bridge for the consumers.
 * */
class StoreAccessor {
  store: Store | null = null;

  setStore = (store: Store | null) => {
    this.store = store;
  };

  getStore = () => {
    return this.store;
  };
}

const singleton = new StoreAccessor();

export default singleton;
