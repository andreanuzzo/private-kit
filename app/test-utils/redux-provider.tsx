// test-utils.js
import React, { ReactElement, ReactNode } from 'react';
import {
  render as rtlRender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react-native';
import { Store, createStore } from 'redux';
import { Provider, DefaultRootState } from 'react-redux';
import rootReducer from '../store/reducers/rootReducer';

type ReduxRenderOptions = RenderOptions & {
  initialState: DefaultRootState;
  store?: Store;
};

type ReduxProviderProps = {
  children?: ReactNode;
};

function render(
  ui: ReactElement<unknown>,
  {
    initialState,
    store = createStore(rootReducer, initialState),
    ...rest
  }: ReduxRenderOptions,
): RenderResult {
  function ReduxProvider({ children }: ReduxProviderProps) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: ReduxProvider, ...rest });
}

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { render };
