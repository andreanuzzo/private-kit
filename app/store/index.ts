import AsyncStorage from '@react-native-community/async-storage';
import { createImmutableStateInvariantMiddleware } from '@reduxjs/toolkit';
import { applyMiddleware, createStore } from 'redux';
import { createMigrate, persistReducer, persistStore } from 'redux-persist';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';

import StoreAccessor from './StoreAccessor';
import onChangedSelectedHealthAuthorities from './middleware/onChangedSelectedHealthAuthorities';

import migrations from './migrations';
import rootReducer from './reducers/rootReducer';

const enhancers = composeWithDevTools(
  applyMiddleware(
    thunk,
    createImmutableStateInvariantMiddleware(),
    onChangedSelectedHealthAuthorities,
  ),
);

export const STORE_VERSION = 2;
const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
  version: STORE_VERSION,
  // add keys of reducers to ignore:
  blacklist: [],
  migrate: createMigrate(migrations),
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = createStore(persistedReducer, {}, enhancers);
StoreAccessor.setStore(store);
export const persistor = persistStore(store);
