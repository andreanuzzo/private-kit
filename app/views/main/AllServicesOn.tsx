import React, { useEffect } from 'react';
import {
  Dimensions,
  ImageBackground,
  StatusBar,
  View,
  TouchableOpacity,
} from 'react-native';
import { useTranslation } from 'react-i18next';
import Pulse from 'react-native-pulse';
import { SvgXml } from 'react-native-svg';

import { Icons, Images } from '../../assets';
import { Typography } from '../../components/Typography';

import { styles } from './style';
import { Colors, Spacing } from '../../styles';
import { Stacks, NavigationProp } from '../../navigation';
import { useSelector, useDispatch } from 'react-redux';
import { RootState } from '../../store/types';
import toggleAutoSubscriptionBanner from '../../store/actions/healthcareAuthorities/toggleAutoSubscriptionBannerAction';

type AllServicesOnProps = {
  navigation: NavigationProp;
};

export const AllServicesOnScreen = ({
  navigation,
}: AllServicesOnProps): JSX.Element => {
  const size = Dimensions.get('window').height;
  const { t } = useTranslation();

  const { autoSubscription, selectedAuthorities } = useSelector(
    (state: RootState) => state.healthcareAuthorities,
  );

  const dispatch = useDispatch();
  useEffect(() => {
    const listener = navigation.addListener('blur', () => {
      if (
        !autoSubscription.bannerDismissed &&
        autoSubscription.selectedAuthority
      ) {
        dispatch(toggleAutoSubscriptionBanner({ overrideValue: false }));
      }
    });
    return listener.remove;
  }, [navigation, dispatch, autoSubscription]);

  const { bannerDismissed, selectedAuthority } = autoSubscription;

  return (
    <View style={{ flex: 1 }}>
      <ImageBackground
        source={Images.BlueGradientBackground}
        style={styles.backgroundImage}>
        <StatusBar
          barStyle='light-content'
          backgroundColor='transparent'
          translucent
        />
        <View style={styles.pulseContainer}>
          <Pulse
            image={{ exportImage: Images.Export }}
            color={Colors.lightestGray}
            numPulses={3}
            diameter={400}
            speed={20}
            duration={2000}
          />
          <SvgXml
            xml={Icons.StateNoContact}
            width={size ? size : 80}
            height={size ? size : 80}
          />
        </View>
      </ImageBackground>

      <View style={styles.mainContainer}>
        <View style={styles.contentAbovePulse} />
        <View style={styles.contentBelowPulse}>
          <Typography style={styles.mainTextBelow}>
            {t('home.gps.all_services_on_header')}
          </Typography>
          <Typography style={styles.subheaderText}>
            {t('home.gps.all_services_on_subheader')}
          </Typography>
          {selectedAuthorities.length === 0 && (
            <Typography style={styles.subheaderText}>
              {t('home.gps.all_services_on_no_ha_available')}
            </Typography>
          )}
        </View>
      </View>

      {!bannerDismissed && !!selectedAuthority && (
        <TouchableOpacity
          style={styles.bottomSheet}
          onPress={() => navigation.navigate(Stacks.Partners)}>
          <View style={{ flex: 1, paddingRight: Spacing.medium }}>
            <Typography
              style={{
                color: Colors.primaryViolet,
              }}>
              <>
                {t('home.gps.auto_subscribe_text', {
                  healthAuthorityName: selectedAuthority.name,
                })}
                <Typography style={styles.hyperlink}>
                  {t('home.gps.auto_subscribe_link_text')}
                </Typography>
              </>
            </Typography>
          </View>
          <SvgXml xml={Icons.ChevronRight} />
        </TouchableOpacity>
      )}
    </View>
  );
};
