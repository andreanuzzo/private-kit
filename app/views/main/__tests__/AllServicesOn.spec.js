import { cleanup, render } from '../../../test-utils/redux-provider';
import React from 'react';

import { AllServicesOnScreen } from '../AllServicesOn';
import { FeatureFlagOption } from '../../../store/types';

afterEach(cleanup);

jest.mock('react-native-pulse');

const navigationMock = {
  addListener: () => ({
    remove: () => {},
  }),
  navigate: () => {},
};

it('all services on matches snapshot', () => {
  const { asJSON } = render(
    <AllServicesOnScreen navigation={navigationMock} />,
    {
      initialState: {
        featureFlags: { flags: { [FeatureFlagOption.MOCK_EXPOSURE]: false } },
      },
    },
  );

  expect(asJSON()).toMatchSnapshot();
});
