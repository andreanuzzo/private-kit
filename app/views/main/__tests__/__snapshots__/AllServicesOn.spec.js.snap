// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all services on matches snapshot 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        Object {
          "flex": 1,
          "height": "110%",
          "justifyContent": "flex-end",
          "resizeMode": "cover",
          "top": 0,
          "width": "100%",
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../app/assets/images/blueGradientBackground.png",
          }
        }
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "height": "110%",
              "width": "100%",
            },
            undefined,
          ]
        }
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "left": 0,
            "position": "absolute",
            "resizeMode": "contain",
            "right": 0,
            "top": "-17%",
          }
        }
      >
        <SvgXml
          height={1334}
          width={1334}
          xml="<svg width=\\"850\\" height=\\"850\\" viewBox=\\"0 0 850 850\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\">
<g opacity=\\"0.2\\">
<circle cx=\\"425\\" cy=\\"425\\" r=\\"85.5\\" stroke=\\"white\\"/>
<circle cx=\\"425\\" cy=\\"425\\" r=\\"149.5\\" stroke=\\"white\\"/>
<circle cx=\\"425\\" cy=\\"425\\" r=\\"217.5\\" stroke=\\"white\\"/>
<circle cx=\\"425\\" cy=\\"425\\" r=\\"283.5\\" stroke=\\"white\\"/>
<circle cx=\\"425\\" cy=\\"425\\" r=\\"352.5\\" stroke=\\"white\\"/>
<circle cx=\\"425\\" cy=\\"425\\" r=\\"424.5\\" stroke=\\"white\\"/>
</g>
<path d=\\"M425 465C447.091 465 465 447.091 465 425C465 402.909 447.091 385 425 385C402.909 385 385 402.909 385 425C385 447.091 402.909 465 425 465Z\\" fill=\\"white\\"/>
<path d=\\"M425 457C442.673 457 457 442.673 457 425C457 407.327 442.673 393 425 393C407.327 393 393 407.327 393 425C393 442.673 407.327 457 425 457Z\\" fill=\\"url(#paint0_linear)\\"/>
<path d=\\"M425 412C417.832 412 412 417.832 412 425C412 432.168 417.832 438 425 438C432.168 438 438 432.168 438 425C438 417.832 432.168 412 425 412ZM425 435.075C419.436 435.075 414.925 430.565 414.925 425C414.925 419.436 419.436 414.925 425 414.925C430.564 414.925 435.075 419.436 435.075 425C435.075 430.565 430.564 435.075 425 435.075Z\\" fill=\\"white\\"/>
<path d=\\"M430.844 419.727C430.219 419.215 429.298 419.305 428.786 419.93L423.15 426.797L421.24 424.334C420.745 423.695 419.826 423.58 419.188 424.074C418.549 424.57 418.433 425.488 418.928 426.127L421.962 430.039C422.234 430.391 422.653 430.599 423.097 430.605C423.104 430.606 423.111 430.606 423.118 430.606C423.556 430.606 423.97 430.409 424.248 430.071L431.046 421.785C431.559 421.161 431.468 420.239 430.844 419.727Z\\" fill=\\"white\\"/>
<defs>
<linearGradient id=\\"paint0_linear\\" x1=\\"425\\" y1=\\"361\\" x2=\\"361\\" y2=\\"425\\" gradientUnits=\\"userSpaceOnUse\\">
<stop stop-color=\\"#6979F8\\"/>
<stop offset=\\"1\\" stop-color=\\"#3A4CD7\\"/>
</linearGradient>
</defs>
</svg>
"
        />
      </View>
    </View>
    <View
      style={
        Object {
          "height": "100%",
          "left": 0,
          "paddingBottom": 12,
          "paddingHorizontal": 24,
          "position": "absolute",
          "right": 0,
          "top": "-10%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "flex-end",
            "paddingBottom": 40,
          }
        }
      />
      <View
        style={
          Object {
            "flex": 1,
            "justifyContent": "flex-start",
            "paddingTop": 80,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#2e2e2e",
                "fontFamily": "IBMPlexSans",
                "fontSize": 17,
                "lineHeight": 24,
              },
              Object {
                "writingDirection": "ltr",
              },
              Object {
                "color": "#ffffff",
                "fontSize": 26,
                "lineHeight": 34,
                "marginBottom": 24,
                "textAlign": "center",
              },
            ]
          }
        >
          PathCheck
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#2e2e2e",
                "fontFamily": "IBMPlexSans",
                "fontSize": 17,
                "lineHeight": 24,
              },
              Object {
                "writingDirection": "ltr",
              },
              Object {
                "color": "#ffffff",
                "fontSize": 18,
                "lineHeight": 24.5,
                "marginBottom": 24,
                "textAlign": "center",
              },
            ]
          }
        >
          PathCheck is saving the places you visit to create your private location diary
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#2e2e2e",
                "fontFamily": "IBMPlexSans",
                "fontSize": 17,
                "lineHeight": 24,
              },
              Object {
                "writingDirection": "ltr",
              },
              Object {
                "color": "#ffffff",
                "fontSize": 18,
                "lineHeight": 24.5,
                "marginBottom": 24,
                "textAlign": "center",
              },
            ]
          }
        >
          You will not receive exposure notifications while you are outside a region served by Health Department partners
        </Text>
      </View>
    </View>
  </View>
</View>
`;
